### PR TITLE
fix(cloudstorage): normalize dav:// and davs:// URL schemes to http(s)://

### PR DESCRIPTION
**Pull request by dani84bs**
### Problem
When users configure a WebDAV storage provider using the custom `dav://` or `davs://` schemes, the backend HTTP request client (`luasocket`'s `socket.http`) fails or crashes because it only natively supports `http://` and `https://` schemes.

### Solution
This PR introduces URL normalization to rewrite standard `dav://` and `davs://` schemes to `http://` and `https://` respectively before executing requests, while leaving conventional `http://` and `https://` URLs untouched.

Key changes:
1. **Frontend / CloudStorage API**:
   - Added `WebDavApi:normalizeUrl(url)` in `frontend/apps/cloudstorage/webdavapi.lua` to map `dav://` -> `http://` and `davs://` -> `https://`.
   - Applied URL normalization across WebDAV API methods (`getJoinedPath`, `listFolder`, `downloadFile`, `uploadFile`, `createFolder`) and in the UI settings dialog callback in `frontend/apps/cloudstorage/webdav.lua`.

2. **CloudStorage Plugin / Provider**:
   - Added a local helper `normalizeUrl(url)` in `plugins/cloudstorage.koplugin/providers/webdav.lua`.
   - Used it in WebDAV provider methods (`getJoinedPath`, `listFolder`) and when saving configuration settings.

3. **Tests**:
   - Created `spec/unit/webdav_spec.lua` with mock-based unit tests for both the frontend API and the plugin provider to verify correct mapping and prevent future regressions.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/15506)

### CODE REVIEW

**Comment by hius07** [comment]
> When users configure a WebDAV storage provider using the custom `dav://` or `davs://` schemes

What are the reasons to do that?

**Comment by dani84bs** [comment]
> > When users configure a WebDAV storage provider using the custom `dav://` or `davs://` schemes
> 
> What are the reasons to do that?

Hi 😁
It's not registered as an official IANA protocol but I believe it's used by some Gnome applications, so some users are copying that address. I helped some users on Reddit configuring nextcloud and some of them were complaining a crash. I told them to change the protocol, but since this is happening quite often I thought that "fixing" this upstream may make KOReader more user friendly

**Comment by hius07** [comment]
Okay.

File: plugins/cloudstorage.koplugin/providers/webdav.lua [hunk 1]
```diff
@@ -316,7 +326,7 @@ The start folder is appended to the server path.]])
                             fields[5] = fields[5]:gsub("/$", "")
                         end
                         item.name     = fields[1]
-                        item.address  = fields[2]
+                        item.address  = normalizeUrl(fields[2])
```
Comment (hius07) [inline plugins/cloudstorage.koplugin/providers/webdav.lua]: So when a user opens "Server settings" again, he will see `http` instead of his original `dav`.
Is it needed? Isn't normalizing on the fly in `getJoinedPath()` and `listFolder()` enough?

Comment (dani84bs) [reply plugins/cloudstorage.koplugin/providers/webdav.lua]: Yes it should be totally enough, I started patching the settings, then I thought that it won't be enough for plugins using the APIs so I added the on the fly patch.
Do you want me to remove the settings one?

Comment (hius07) [reply plugins/cloudstorage.koplugin/providers/webdav.lua]: I think so, letting a user be happy with his address.

Comment (dani84bs) [reply plugins/cloudstorage.koplugin/providers/webdav.lua]: Ok, I'll be back with a commit removing that part

Comment (dani84bs) [reply plugins/cloudstorage.koplugin/providers/webdav.lua]: Ok, done 😀

**Review by hius07 — APPROVED** [review]
Thanks.

**Comment by dani84bs** [comment]
> Thanks.

Thank you!

**Comment by Frenzie** [comment]
Crash even? Did you add a pcall or some such in that case? Not so much for this special case but for a typo like htp or something.

**Comment by dani84bs** [comment]
> Crash even? Did you add a pcall or some such in that case? Not so much for this special case but for a typo like htp or something.

Yes, they reported crashes, I reproduced only on my Mac and it crashed as well. Actually I thought about the usecase but I supposed that the pcall was missing for a reason 😅

 If you want tomorrow I'll create more edge cases tests, trace the crash and wrap it in a pcall.

**Comment by hius07** [comment]
The app will be removed soon, no need to fix it I believe.
The plugin is sandboxed.

**Comment by Frenzie** [comment]
Alright.

**Comment by dani84bs** [comment]
> Alright.

Hi @Frenzie I pushed a commit where I wrap the webdav functions in pcall. As for @hius07 comment I fixed it only in the plugin and left the service as is, since is going to be dismissed soon.
Thank you for your work 😀
